### ISSUE TITLE
[Bug] WAM interactive dialog (the Account Picker) is immediately canceled when shown in an elevated app

### ISSUE THREAD

**Issue by BenPryor** [AzureAD/microsoft-authentication-library-for-dotnet#2560]
**Which Version of MSAL are you using ?**
MSAL 4.29.0

**Platform**
.NET 5

**What authentication flow has the issue?**
Desktop / Mobile: Interactive

**Is this a new or existing app?**
This is a new app or experiment

**Repro**
C# console app, .NET 5, target framework `net5-windows10.0.17763.0`.
```csharp
static async Task Main()
{
    IPublicClientApplication app = PublicClientApplicationBuilder
        .Create("my-client-id")
        .WithAuthority("https://login.microsoftonline.com/my-tenant-id")
        .WithExperimentalFeatures()
        .WithBroker()
        .Build();

    await app.AcquireTokenInteractive(new string[] { "my-resource-id/.default" })
        .ExecuteAsync();
}
```

**Expected behavior**
The WAM interactive dialog (account picker) should be shown.

**Actual behavior**
If the console app is run elevated, the WAM dialog briefly flashes and is immediately canceled (MSAL returns `authentication_canceled`). If the console app is run non-elevated, the WAM dialog is shown as expected.

**Additional context/ Logs / Screenshots**
I see this issue on .NET framework, .NET Core 3.1, and .NET 5.

In case this is a bug in WAM itself, I am reproing this on Win10 20H2.

**Comment by bgavrilMS** [comment]
A workaround has been provided by the Windows team: 

```
This is one of the native APIs that is historically known to have problems calling from .Net. The issue is that this needs to be called before COM remoting is initialized in a process or it will happen implicitly, and there are things CLR may do early in the process lifetime that cause this. This is mentioned (with some details inaccurate) on the pinvoke.net page:
You shouldn't call CoInitializeSecurity from managed code.  http://pinvoke.net/default.aspx/ole32/CoInitializeSecurity.html 

Most of what you can do with CoInitializeSecurity, including access permissions as needed here, can also be done with an AppID key for the exe (see [here ](https://docs.microsoft.com/en-us/windows/win32/com/appid) and [here](https://docs.microsoft.com/en-us/windows/win32/com/appid-key)). Note that for COM to read this from an elevated process it must be in HKLM.
```

**Comment by mjcheetham** [comment]
One possible interesting solution may be to set the COM security settings _on the specific proxy/interface_ for the WAM account control, rather than try and change the COM security settings for the entire process via the registry or `CoInitializeSecurity`.

> Another common scenario in which you might set process-wide security programmatically is when you would like to set default security for the entire process **but you have one or more objects within that process that expose interfaces with special security requirements.** In this case, you can call CoInitializeSecurity to set security for the process, allowing COM to handle most of the security checks, **and you can call other methods to set security for the objects with special security needs.** Calling these methods and functions is described in [Setting Security at the Interface Proxy Level](https://docs.microsoft.com/en-us/windows/win32/com/setting-security-at-the-interface-proxy-level).

Source: https://docs.microsoft.com/en-us/windows/win32/com/setting-processwide-security-with-coinitializesecurity

> Sometimes the client needs fine-grained control over the security on calls to particular interfaces. For example, security might be set at a low level for the process but calls to a particular interface might require a higher authentication level, such as encryption. The methods of the IClientSecurity interface allow the client to change the security settings associated with calls to a particular interface by controlling the security settings at the interface-proxy level.
> [...]
> [SetBlanket](https://docs.microsoft.com/en-us/windows/win32/api/objidl/nf-objidl-iclientsecurity-setblanket) is commonly used to raise the authentication level for a particular interface proxy to a higher level of security protection. **However, in some situations, it might also be helpful to lower the authentication level for a particular interface proxy.** For instance, suppose the default authentication level for the process is some value other than RPC_C_AUTHN_LEVEL_NONE and the client and server are in separate domains that do not trust each other. In this case, calls to the server will fail unless the client calls SetBlanket to lower the authentication level to RPC_C_AUTHN_LEVEL_NONE.

Source: https://docs.microsoft.com/en-us/windows/win32/com/setting-security-at-the-interface-proxy-level

This sounds like what's happening, no?

> "For instance, suppose the default authentication level for the process is some value other than RPC_C_AUTHN_LEVEL_NONE and the client and server are in separate domains that do not trust each other."

Not sure if you've explored this possibility? @bgavrilMS @pmaytak

**Comment by mjcheetham** [comment]
From Git Credential Manager, we're looking to try calling the `CoInitializeSecurity` function early in our process' lifetime, with appropriate error handling and user docs/links.

https://github.com/microsoft/Git-Credential-Manager-Core/pull/344

If a better solution appears, we'll revisit and see if that provides a better UX.

**Comment by pmaytak** [comment]
Yep, seems like `CoInitializeSecurity` is the appropriate workaround for now. And setting the security for an interface proxy, unfortunately, doesn't affect the callee process calling back into the elevated caller process. Short-term we'll probably just make the error message in MSAL.NET more descriptive, until a fix exists.

**Comment by pmaytak** [comment]
Starting in next MSAL version (4.32), the error will be a bit more descriptive when app process is elevated and WAM returns no accounts. If the app must be run in admin mode, in short-term the recommendation is to call a helper method, `WindowsNativeUtils.InitializeProcessSecurity`. Described a bit more in [Using WAM in an elevated process](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/WAM#using-wam-in-an-elevated-process) wiki.

**Comment by techyrajeev** [comment]
Hi Team,
When are we expecting this fix to be available with MSAL.NET?

**Comment by bgavrilMS** [comment]
@techyrajeev - please see the workaround @pmaytak  posted, this should unblock you. We are not planning to fix this in MSAL 4.x, but as part of a larger move to MSAL 5.x

**Comment by bgavrilMS** [comment]
@pmaytak - did you try this workaround from WinForms or WPF on .NET Classic ? I cannot get this to work :(. The error is "called to late". This is the problem @techyrajeev  is running into as well.

**Comment by pmaytak** [comment]
Hmm, I think I did but I don't remember exactly. I faintly remember that I was trying to repro this error but wasn't able to. (Perhaps I only tried .NET Core.) But yea, that's a valid error - means .NET itself called security initialization code first.

**Comment by bgavrilMS** [comment]
Fix proposal: 

- MSAL should fallback to a webview if the app is running as admin and the COM security stuff is not configured.

**Comment by lacomc** [comment]
I'm trying to use the provided workaround (WindowsNativeUtils.InitializeProcessSecurity()) to enable WAM sign-in from an elevated process, and I'm getting a very generic error message:

> "Failure setting process security to enable WAM Account Picker in an elevated process (0x80010119). For troubleshooting details, see https://aka.ms/msal-net-wam ."

https://aka.ms/msal-net-wam doesn't have any details on running in an elevated process, and neither does the link dropped above (https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/WAM#using-wam-in-an-elevated-process). Is there troubleshooting information on this anywhere? Or code samples so I can understand if my usage is incorrect?

**Comment by pmaytak** [comment]
@lacomc We are currently moving to a new better WAM experience (currently in preview), and the wiki was updated for that. The current (soon-to-be-legacy) WAM experience is documented [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/scenario-desktop-acquire-token-wam#either-the-user-canceled-the-authentication-or-the-wam-account-picker-crashed-because-the-app-is-running-in-an-elevated-process-error-message). Unfortunately this workaround doesn't all the time - sometimes the security is set before the app code runs, which results in an error.

I suggest trying the new WAM which works fine in elevated processes. [This](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/wam#new-wam-preview-in-msal-444) describes how to implement.

Also I see that you're in a Visual Studio team? CC'ing @gladjohn who has been collaborating with the VS team on testing the preview WAM package.

**Comment by bgavrilMS** [comment]
I'm going to close this as "fixed" with the newer broker implementation. See previous comment for details.

**Comment by maheshdevaraju** [comment]
I'm using WAM broker(MSAL.NET 4.56.0) in a VSTO Excel Add-in application based on .NET Framework 4.8. 

The app works without any issues when it is not elevated, but when the App is elevated I face the "authentication_canceled" error code.

```c#
static async Task Main()
{
    IPublicClientApplication app = PublicClientApplicationBuilder
        .Create("CLIENT-ID")
        .WithDefaultRedirectUri()
        .WithBroker(new BrokerOptions(BrokerOptions.OperatingSystems.Windows))
        .Build();

    await app.AcquireTokenInteractive(scopes)
                        .WithParentActivityOrWindow(hWnd)
                        .ExecuteAsync()
}
```
Based upon the comments above I tried to use WindowsNativeUtils.InitializeProcessSecurity, but looks like this is deprecated. Is there any other way I can fix this issue?

**Comment by pmaytak** [comment]
@maheshdevaraju 4.56.0 uses a new WAM broker implementation that works with elevated privileges. So it's a different issue from this original one. Please create a separate issue with repro steps, details, and verbose logs and we'll look into it.

**Comment by maheshdevaraju** [comment]
New issue has been created. #4360
